sympy/codegen/algorithms.py

```diff
@@ -0,0 +1,125 @@
+from __future__ import (absolute_import, division, print_function)
+
+from sympy import And, Gt, Lt, Abs, Dummy, oo, Tuple, Symbol, Function, Pow
+from sympy.codegen.ast import (
+    Assignment, AddAugmentedAssignment, CodeBlock, Declaration, FunctionDefinition,
+    Print, Return, Scope, Statement, While, Variable, Pointer, real
+)
+
+""" This module collects functions for constructing ASTs representing algorithms. """
+
+def newtons_method(expr, wrt, atol=1e-12, delta=None, debug=False,
+                   itermax=None, counter=None):
+    """ Generates an AST for Newton-Raphson method (a root-finding algorithm).
+
+    Returns an abstract syntax tree (AST) based on ``sympy.codegen.ast`` for Netwon's
+    method of root-finding.
+
+    Parameters
+    ----------
+    expr : expression
+    wrt : Symbol
+        With respect to, i.e. what is the variable.
+    atol : number or expr
+        Absolute tolerance (stopping criterion)
+    delta : Symbol
+        Will be a ``Dummy`` if ``None``.
+    debug : bool
+        Whether to print convergence information during iterations
+    itermax : number or expr
+        Maximum number of iterations.
+    counter : Symbol
+        Will be a ``Dummy`` if ``None``.
+
+    Examples
+    --------
+    >>> from sympy import symbols, cos
+    >>> from sympy.codegen.ast import Assignment
+    >>> from sympy.codegen.algorithms import newtons_method
+    >>> x, dx, atol = symbols('x dx atol')
+    >>> expr = cos(x) - x**3
+    >>> algo = newtons_method(expr, x, atol, dx)
+    >>> algo.has(Assignment(dx, -expr/expr.diff(x)))
+    True
+
+    References
+    ==========
+    https://en.wikipedia.org/wiki/Newton%27s_method
```

Comment:
not sure if a blank line is needed here -- see rendering and/or check other files